README.md
[![da](https://flagcdn.com/w20/dk.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.da.md)
[![de](https://flagcdn.com/w20/de.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.de.md)
[![fr](https://flagcdn.com/w20/fr.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.fr.md)
[![it](https://flagcdn.com/w20/it.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.it.md)
[![jp](https://flagcdn.com/w20/jp.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.jp.md)
[![nl](https://flagcdn.com/w20/nl.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.nl.md)
[![pl](https://flagcdn.com/w20/pl.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.pl.md)
[![pt-PT](https://flagcdn.com/w20/pt.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.pt-PT.md)
[![pt-BR](https://flagcdn.com/w20/br.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.pt-BR.md)
[![ru](https://flagcdn.com/w20/ru.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.ru.md)
[![sp](https://flagcdn.com/w20/es.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.sp.md)
[![sv](https://flagcdn.com/w20/sv.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.sv.md)
[![tr](https://flagcdn.com/w20/tr.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.tr.md)
[![zh](https://flagcdn.com/w20/cn.png)](https://github.com/schmidhorst/synology-autorun/blob/main/README.zh.md)

# autorun package V1.10 for Synology NAS
Executes a script when connecting external storages (USB / eSATA) to a Synology NAS with DSM 7.x. Typical use is to copy or backup some files.
In Synologies Task Scheduler there is the possibility to create triggered tasks. But for the trigger event there are only Boot-up and Shutdown available. There are no USB events like 'external storage mounted' available. This deficit is compensated by this tool.

During the package installation to your NAS you have to enter the name of the shell script, which should be executed at the trigger event 'external storage device mounted'. The script can be located either in the root folder of the external device or internal on your NAS. Optionally you can exject that external device after your script is finished. A security option is to register the SHA256 checksum of the executed script and to restrict further executions to scripts matching the previously registered checksums.

## [License](https://htmlpreview.github.io/?https://github.com/schmidhorst/synology-autorun/blob/main/package/ui/licence_enu.html)

## Disclaimer and Issue Tracker
You are using everything here at your own risk.
For issues please use the [issue tracker](https://github.com/schmidhorst/synology-autorun/issues) with German or English language

## Build and Installation
* If you want to build the package yourself from the source: Download all files. Set the file 'build' to executable and run that script to generate the *.spk file. No toolkit needed.
* Or download the *.spk file from ["Releases"](https://github.com/schmidhorst/synology-autorun/releases), "Assets" to your computer.

Then use "Manual Install" in the Package Center.

Third Party packages are restricted by Synology in DSM 7. Since autorun does require root permission to perform its job an additional manual step is required after the installation before a "Run" can be done successfully.

SSH to your NAS (as an admin user) and execute the following command:
```shell
sudo cp /var/packages/autorun/conf/privilege.root /var/packages/autorun/conf/privilege
```
Alternative to SSH:
Go to Control Panel => Task Scheduler => Create => Scheduled Task => User-defined Script. In the "General" tab set any task name, select 'root' as user. In the "Task Settings" tab enter
```shell
cp /var/packages/autorun/conf/privilege.root /var/packages/autorun/conf/privilege
```
as "Run command". Finish it with OK. When you are requested to execute that command now during package installation, then go to the task scheduler, select that task and "Run" it.

Under https://www.cphub.net/ in the Package Center there are [elder versions](https://github.com/reidemei/synology-autorun) for elder DSM versions available:
* DSM 7: 1.8
* DSM 6: 1.7
* DSM 5: 1.6
* elder: 1.3

## Credits and References
- Thanks to [Jan Reidemeister](https://github.com/reidemei) for his [Version 1.8](https://github.com/reidemei/synology-autorun) and his [License](https://github.com/reidemei/synology-autorun/blob/main/LICENSE)
- Thanks to the [Synology Forum Thread about that autorun package](https://www.synology-forum.de/threads/autorun-fuer-ext-datentraeger.18360/)
- Thanks to [toafez Tommes](https://github.com/toafez) and his [Demo Package](https://github.com/toafez/DSM7DemoSPK)
- Thanks to  [geimist Stephan Geisler](https://github.com/geimist) and his tip to use the [DeepL API](https://www.deepl.com/docs-api) for translations to other languages.
- Thanks to QTip and his explanation about the help integration [Integration einer Hilfe in DSM 5.1-](https://www.synology-wiki.de/index.php/Integration_einer_Hilfe_in_DSM_5.1-)



CHANGELOG
 !!! A MANUAL STEP IS REQUIRED AFTER THE INSTALLATION !!!  See V1.8 below!
1.10.0011
 - Eject can be triggered by a list of exit codes, no more only at exit code 100
 - Help added, integrated to the Synology Help
 - Several improvements (Installation Wizzard, Logging, ...)
 - Several Bugs fixed (LED control, quoting in strings, ...)

1.10.0 Horst Schmid

 - Eject USB Drive (at exit code 100) reworked, so that it works also on DSM 7.x
 - UDEV daemon is no more blocked if a long running scripts is strarted by this tool
 - Allow to use the status LED and/or the copy LED
 - Debug logfile for installation/upgrade actions moved from @appstore (target) to /var/tmp/autorun.log folder
 - Logfile for drive connect and drive eject actions added (in @appdata / var folder)
 - Config files moved from @appstore (target) to @appdata (var) folder
 - At upgrade the WIZARD is used like during the 1st installation, but the presets are read from actual configuration
 - Ability to restrict the allowed scripts on the USB drives to scripts with pre-registered hash values. (similar to heandr V1.9)
 - In case of using exit code 100 to eject the USB drive now a further script can be called afterwards e.g. to power down the USB drive (if eject was successfully)
 - Code for local Backup from V1.6, which was since V1.7 no more working, removed
 - A new GUI added
 - Desktop notification implemented again
 - Internationalization done with machine tranlations via DeepL
 - The script, which is started after external drive mounting can now be located either on the external drive or internally.

 - note: for DSM 6.x use the version 1.7, which can be found
   at https://github.com/reidemei/synology-autorun/releases/tag/v1.7


1.8.x/1.9
  Forks from Jip-Hop (https://github.com/Jip-Hop/synology-autorun) and JC-23 (https://github.com/JC-23/synology-autorun)
  added: optional feature to verify sha256 fingerprints of scripts before running them

1.8
 !!! A MANUAL STEP IS REQUIRED AFTER THE INSTALLATION !!!
 Third Party packages are restricted by Synology in DSM 7. Since autorun does require root
 permission to perform its job an additional manual step is required after the installation.
 SSH to your NAS (as an admin user) and execute the following command:
   sudo cp /var/packages/autorun/conf/privilege.root /var/packages/autorun/conf/privilege

 - changed: compatibility with DSM 7
 - removed: desktop notification
 - note: for DSM 6.x use the version 1.7, which can be found
   at https://github.com/reidemei/synology-autorun/releases/tag/v1.7

1.7
 - changed: compatibility with DSM 6
 - changed: configuration moved to setup, just un- and reinstall if you want to change something
 - fixed: eject does not properly work sometimes
 - removed: all GUI, this includes the local backup scripts
            (sorry, but I'm tired of the constant changes by Synology)
 - note: for DSM 5.x use the version 1.6, which can be found
   at https://github.com/reidemei/synology-autorun/releases/tag/v1.6

1.6
 - changed: compatibility with DSM 5.1

1.5
 - changed: configurable wait times before starting the script

1.4
 !!! YOU HAVE TO RECREATE THE SCRIPTS FOR THE LOCAL BACKUP !!!

 - changed: requires DSM 5.0+
 - changed: include Perl package as requirement
 - changed: local backup - adjustments for changes in DSM 5.0
 - note: for DSM 4.3 and below use the version 1.3, which can be found
   at https://github.com/reidemei/synology-autorun/releases/tag/v1.3

1.3
 - changed: allow access to settings by any user in the admin group (code [c] dakkar)
 - changed: link to restore encrypted backup added to main page
 - fixed: unable to create backup-script when multiple drives attached (on some browsers)

1.2
 - new: show warning for required Perl package

1.1
 - new: report space left on the device in the log
 - new: optional DSM notification when finished
 - new: mount encrypted backup for restore - option for file name encryption added
 - changed: adjusted notes for file name encryption
 - fixed: script not executed when name changed

1.0
 - new: local backup - option for file name encryption added
 - changed: local backup - name must be entered manually (Synology changes that too often)
 - changed: local backup - 10s sleep added to give the drive some time to be properly mounted

0.9
 - new: now supports update
 - changed: requires DSM 3.2

0.8
 - fixed: don't log passwords

0.7
 - new: local backup - option for encryption added

0.6
 - changed: requires DSM 3.1
 - fixed: local backup - wait till backup is finished

0.5
 - fixed: incorrect handling of DS restarts

0.4
 - new: create scripts for the local backup

0.3
 - new: settings page in DSM added

0.2
 - new: beep at start and finish
 - new: use DS led for status

0.1
 - intial test


package/ui/modules/parse_language.sh
#!/bin/bash
# Filename: parse_language.sh - coded in utf-8
# taken from
#   SPKdevDSM7
#   Copyright (C) 2022 by Tommes
#   Member of the German Synology Community Forum
#     License GNU GPLv3
#   https://www.gnu.org/licenses/gpl-3.0.html

# Adopted to the needs of Autorun by Horst Schmid
#      Copyright (C) 2022...2023 by Horst Schmid

#********************************************************************#
#  Description: Script get the currently used language               #
#               Either the language setup for the logged-in user     #
#               or the language according the web browser setup      #
#               or the setup for the DSM display or DSM messages     #
#  Author 1:    QTip from the german Synology support forum          #
#  Copyright:   2016-2018 by QTip                                    #
#  Author 2:    Modified 2022 by Tommes                              #
#  Author 3:    Horst Schmid, 2022...2023                            #
#  License:     GNU GPLv3                                            #
#  ----------------------------------------------------------------  #
#  Version:     2023-01-06                                           #
#********************************************************************#
bDebugPL=0
DTFMT="+%Y-%m-%d %H:%M:%S"
if [[ -z "$SCRIPT_NAME" ]]; then  # direct start in debug run
  SCRIPTPATHTHISpl="$( cd -- "$(/bin/dirname "${BASH_SOURCE[0]}")" >/dev/null 2>&1 ; /bin/pwd -P )" # e.g. /volumeX/@appstore/<app>/ui
  scriptpathParent=${SCRIPTPATHTHISpl%/*}
  cd "$scriptpathParent"
  app_name="autorun"  # may be needed to change if appName is changed!
  SCRIPT_NAME="/webman/3rdparty/${app_name}"
  echo "parse_language.sh started and switched to debug mode ..."
  bDebugPL=1
  source "$SCRIPTPATHTHISpl/parse_hlp.sh"
else
  app_link=${SCRIPT_NAME%/*} # "/webman/3rdparty/<appName>"
  app_name=${app_link##*/} # "<appName>"
fi

LOG="/var/tmp/${app_name}.log"
logInfoNoEcho 6 "parse_language.sh started with param1='$1', whoami=$(whoami) ..."
if [[ "$bDebugPL" -eq "1" ]]; then
  echo "see $LOG"
fi
# # even if we have had ${login_result} != "success" in showlog.cgi, then still no access!
# usersettingsfile="/usr/syno/etc/preference/${user}/usersettings"
# Therefore languages are pre-fetched in the start-stop script:
# DSM language
gui_lang=$(/bin/get_key_value /etc/synoinfo.conf language) # Display language, may be 'def'!
declare -A ISO2SYNO
ISO2SYNO=( ["de"]="ger" ["en"]="enu" ["zh"]="chs" ["cs"]="csy" ["jp"]="jpn" ["ko"]="krn" ["da"]="dan" ["fr"]="fre" ["it"]="ita" ["nl"]="nld" ["no"]="nor" ["pl"]="plk" ["ru"]="rus" ["sp"]="spn" ["sv"]="sve" ["hu"]="hun" ["tr"]="trk" ["pt-BR"]="ptb" ["pt"]="ptg" ["pt-PT"]="ptg" )
declare -A SYNO2ISO # used in *.cgi files
# shellcheck disable=2034
SYNO2ISO=(   ["ger"]="de" ["enu"]="en" ["chs"]="zh" ["csy"]="cs" ["jpn"]="jp" ["krn"]="ko" ["dan"]="da" ["fre"]="fr" ["ita"]="it" ["nld"]="nl" ["nor"]="no" ["plk"]="pl" ["rus"]="ru" ["spn"]="sp" ["sve"]="sv" ["hun"]="hu" ["trk"]="tr" ["ptg"]="pt"  ["ptb"]="pt-BR" )
# SYNO2ISO=( ["ger"]="de" ["enu"]="en" ["chs"]="zh" ["csy"]="cs" ["jpn"]="ja" ["krn"]="ko" ["dan"]="da" ["fre"]="fr" ["ita"]="it" ["nld"]="nl"              ["plk"]="pl" ["rus"]="ru" ["spn"]="es" ["sve"]="sv" ["hun"]="hu" ["trk"]="tr" ["ptb"]="pt-BR" ["ptg"]="pt-PT" )
# Japan: ja (FIPS 10 = U.S. Federal Information Processing Standard No. 10) or jp (ISO 3166-1)?

if [[ -n "$SYNOPKG_DSM_LANGUAGE" ]]; then # Language of the logged-in user, in cgi files not available
  lngDsmUser="$SYNOPKG_DSM_LANGUAGE"
fi

if [[ -n "${LANG}" ]]; then # in cgi files not set
  logInfoNoEcho 8 "env LANG='$LANG'"
  env_lng="${LANG:0:2}"
  env_lng=${ISO2SYNO[$env_lng]}
fi

lngDsm2=$(/bin/get_key_value "/etc/synoinfo.conf" "language") # Display Language, may be "def"
lngMail=$(/bin/get_key_value "/etc/synoinfo.conf" "maillang") # Notification Language, e.g. ger, global setting, not individual user!

httpSynLngs=""
if [ -n "${HTTP_ACCEPT_LANGUAGE}" ] ; then  # WebBrowser-Preset available in cgi files, e.g. 'de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7'
  # try to translate the ISO language codes now to SYNO language codes:
  # logInfoNoEcho 6 "From web browser regional settings: HTTP_ACCEPT_LANGUAGE='${HTTP_ACCEPT_LANGUAGE}'"
  if [[ -n "${HTTP_ACCEPT_LANGUAGE}" ]]; then
    # mapfile -d "," -t httpLngs <<< "${HTTP_ACCEPT_LANGUAGE}" # here-string <<< appends a newline!
    mapfile -d "," -t httpLngs < <(/bin/printf '%s' "$HTTP_ACCEPT_LANGUAGE") # process substitution should be available also in ash with bash-compatibility
    # httpLngs[-1]=$(echo "${httpLngs[-1]}" | sed -z 's|\n$||' ) # remove the \n which was appended to last item by "<<<"
    msg1="${#httpLngs[@]} Elments in httpLngs[@]:"
    for ((i=0; i<${#httpLngs[@]}; i+=1)); do
      b1=${httpLngs[i]%%;*} # remove e.g. ";q=0.7", remaining e.g. "pt-PT, de-DE"
      b2=${httpLngs[i]:0:2}
      b2=${b2,,} # to lower case, should not be necessary
      msg1="$msg1 ${httpLngs[i]}"
      [[ -n ${ISO2SYNO[$b1]} ]] && msg1="$msg1, b1='$b1'==>'${ISO2SYNO[$b1]}"
      [[ -n ${ISO2SYNO[$b1]} ]] && msg1="$msg1, b2='$b2'==>'${ISO2SYNO[$b2]}"
      if [[ -n "${ISO2SYNO[$b1]}" ]] && [[ "$httpSynLngs" != *"${ISO2SYNO[$b1]}"* ]]; then
        httpSynLngs="$httpSynLngs${ISO2SYNO[$b1]} "
      elif [[ -n "${ISO2SYNO[$b2]}" ]] && [[ "$httpSynLngs" != *"${ISO2SYNO[$b2]}"* ]]; then
        httpSynLngs="$httpSynLngs${ISO2SYNO[$b2]} "
      elif [[ -z "${ISO2SYNO[$b1]}" ]] && [[ -z "${ISO2SYNO[$b2]}" ]]; then
        msg1="$msg1: no Syno language"
      fi
    done
    # logInfoNoEcho 6 "From web browser regional settings: HTTP_ACCEPT_LANGUAGE='${HTTP_ACCEPT_LANGUAGE}'<br>$msg1"
  fi
fi

logInfoNoEcho 8 "From web browser regional settings: HTTP_ACCEPT_LANGUAGE='${HTTP_ACCEPT_LANGUAGE}'<br>${msg1}<br>lngDsmUser='$lngDsmUser', httpLng='$httpSynLngs',lngDsm2='$lngDsm2', env_LANG='$env_lng', gui_lang='$gui_lang', lngMail='$lngMail'"
# shellcheck disable=2206
languages=( $lngDsmUser $httpSynLngs $lngDsm2 $env_lng $gui_lang $lngMail )
# logInfoNoEcho 6 "languages with precedence to check for an available translation: ${languages[@]}"
used_lang=""
for lngx in "${languages[@]}"; do
  if [[ -n "$lngx" ]] && [[ "$lngx" != "def" ]]; then
    if [[ -f "texts/${lngx}/lang.txt" ]]; then
      used_lang=$lngx
      break
    else
      logInfoNoEcho 6 "No translation file found for '$lngx'"
    fi
  fi
done
if [[ -z "$used_lang" ]]; then
  used_lang="enu"
  logInfoNoEcho 3 "Language fallback to English as no other selection found"
fi
# shellcheck disable=2034
lngUser=$used_lang
# logInfoNoEcho 6 "Selected Language '$used_lang'"
lngFile=texts/${used_lang}/lang.txt
if [[ ! -f "$lngFile" ]]; then
  logError "File '$lngFile' not available. pwd='$(/bin/pwd)'"
fi
# shellcheck source=../texts/enu/lang.txt
source "$lngFile" # setup $fingerprint0, $fingerprint1count0, $fingerprint1count1, $fingerprint2, ...
res=$?
# logInfoNoEcho 7 "Result from 'source $lngFile' is '$res'"
if [[ "$res" -ne 0 ]]; then
  logInfoNoEcho 1 "Bad result from 'source $lngFile' is '$res'!"
  if [[ "$used_lang" != "enu" ]]; then
  # try fallback to English:
    # shellcheck source=../texts/enu/lang.txt
    source "texts/enu/lang.txt"
    res=$?
    logInfoNoEcho 7 "Result from 'source $lngFile' is '$res'"
    if [[ "$res" -ne 0 ]]; then
      logInfoNoEcho 1 "Failed to fallback to Englisch, result from 'source texts/enu/lang.txt' is '$res'"
      # dsmnotify
    fi # if [[ "$res" -ne 0 ]]
  fi # if [[ "$used_lang" != "enu" ]]
fi # if [[ "$res" -ne 0 ]]
if [[ $bDebugPL -eq 1 ]]; then
  echo "... parse_language.sh done with res=$res"
fi
logInfoNoEcho 6 "... parse_language.sh done, language set to '$used_lang'"
